# bacsnp - Deciphering baculovirus populations

Baculoviruses and other nuclear arthropod-specific large dsDNA viruses (NALDV) have large genomes of up to hundreds of kbp with up to hundreds of open reading frames (ORF). The size of the genome makes it very difficult to analyse the composition of virus populations, since current sequencing techniques cannot (Illumina sequencing) or can only with great difficulty (Nanopore sequencing) sequence entire genomes or significant genome fragments. Genetic markers such as insertion/deletion or single nucleotide polymorphisms (SNP) are usually used for virus population analysis. SNPs are particularly suitable for analysing intra-isolate specific variation, as there are many bioinformatic tools and workflows available and established to detect SNPs based on raw sequencing data (usually in fastq or fastqsanger format).

In virus populations, SNP positions can be used to count the frequency of the occurring nucleotides. In theory, A, T, G and C can occur in any position. The nucleotides can be determined using the sequencing data because not only one genome but a multitude of genomes is sequenced, which reflects the virus population including its genetic variations.

If some SNP positions are specific for a sequenced sample or isolate, then this position and the occuring nucleotide can be used as a marker to find this isolate/sample in other sequences, even if it is only present in a mixture with another type.

The ***bacsnp*** tool was developed to assign specificities to SNP positions. The tool is written in the R programming language and uses the Variant Call Format generated by Mpileup as input.

How the bacsnp can be integrated into a workflow can be found in detail in the following publication:

*Wennmann, Jörg T.; Fan, Jiangbin; Jehle, Johannes A. **2020**. ["Bacsnp: Using Single Nucleotide Polymorphism (SNP) Specificities and Frequencies to Identify Genotype Composition in Baculoviruses."](https://www.mdpi.com/1999-4915/12/6/625) Viruses 12, no. 6: 625.*

## Prerequisites

The tool was written in the R programming language and therefore requires the latest version of R to be installed on the computer. In addition, I recommend using RStudio. The following packages for R should also be installed:

-   devtools
-   ggplot2
-   vcfR

To install the **devtool** and **ggplot2** package, the following steps can be carried out in the R console:

```         
install.packages("devtools")
install.packages("ggplot2")
install.packages("vcfR")
```

Check if everything is installed correctly by starting the packages.

```         
library(devtools)
library(ggplot2)
library(vcfR)
```

The packages should start without any error messages.

### How to install bacsnp from Github

Next we look at how we install bacsnp directly from Github in R. This step requires the **devtools** package, which is available via CRAN. It must be installed and initiated.

```         
library(devtools)
```

The bacsnp package is locaetd in my repository wennj/bacsnp. For its installation you type

```         
install_github("wennj/bacsnp", build_vignettes = TRUE)
```

which should install the bacsnp package along with its vignette (the packages introduction and description). After installation, you should be able to start it as usual to use it straight away.

```         
library(bacsnp)
```

## Getting started

In this section we will focus on how to use bacsnp. Bacsnp requires a Variant Call Format (VCF) file that was created with Mpileup for one or more sequenced isolates of a baculovirus species. The vignette of the package also offers further information, which can be opened with the following command:

```         
vignette("bacsnp")
```

...

## Running the tests

To be continued.

### Break down into end to end tests

To be continued.

### And coding style tests

To be continued.

## Contributing

To be continued.

## Versioning

To be continued.

## Authors

To be continued.

## License

To be continued.

## Acknowledgments

To be continued.
